### PR TITLE
critical : add p2p architecture

### LINKED ISSUE #26: [honeybadger] define transport (grpc) interface
**Subject**

**Detail**

**Reference**

### PR DESCRIPTION
**Pull request by AgwaB**
## Related Issue

resolve: #26 

## Description

- stream.go :  and has networking logic
- conn.go : manage connection (grpc)
- `conn` has a `stream`

## Checklist

Thank you for your contribution.
Before submitting this PR, please make sure:

- [ ] Test case
- [x] End of Work

### CODE REVIEW

File: conn.go [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+type Connection interface {
+	Send(data []byte, typ string, successCallBack func(interface{}), errCallBack func(error))
+	GetIP() string
+	GetID() string
+	Close()
+	Start() error
+	Handle()
+}
+
+type GrpcConnection struct {
```
Comment (zeroFruit) [inline conn.go:38]: I think we need to add `id` field with type `ConnId`

File: conn.go [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+	Conn Connection
+}
+
+type Connection interface {
+	Send(data []byte, typ string, successCallBack func(interface{}), errCallBack func(error))
+	GetIP() string
+	GetID() string
+	Close()
+	Start() error
+	Handle()
```
Comment (zeroFruit) [inline conn.go]: `Handle(handler Handler)` plz

File: conn.go [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+}
+
+type GrpcConnection struct {
+	streamWrapper StreamWrapper
+	outChannel    chan *innerMessage
+	readChannel   chan *pb.Message
+	stopChannel   chan struct{}
```
Comment (zeroFruit) [inline conn.go]: how about Channel -> Chan

File: conn.go [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+}
+
+type Message struct {
+	*pb.Message
+	Conn Connection
+}
+
+type Connection interface {
+	Send(data []byte, typ string, successCallBack func(interface{}), errCallBack func(error))
+	GetIP() string
```
Comment (hihiboss) [inline conn.go]: In Go convention, Get is not used. How about `func IP()`?

File: conn.go [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+	GetID() string
```
Comment (hihiboss) [inline conn.go]: In Go convention, Get is not used. How about `func ID()`?

**Review by hihiboss — COMMENTED** [review]
My comments are just about Go convention :)

File: conn.go (hunk 4 above)
Comment (zeroFruit) [reply conn.go]: `Ip` plz

File: conn.go (hunk 5 above)
Comment (zeroFruit) [reply conn.go]: `Id` plz

File: conn.go [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+	OnSuccess func(interface{})
+}
+
+// message used with other nodes
+type Message struct {
+	*pb.Message
+	Conn Connection
+}
+
+// request handler
```
Comment (zeroFruit) [inline conn.go:24]: plz more specify the comment, can't understand what `Handler` is for

File: conn.go [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+type Handler interface {
+	ServeRequest(msg Message)
+}
+
+type Connection interface {
+	Send(data []byte, typ string, successCallBack func(interface{}), errCallBack func(error))
+	Ip() MemberId
```
Comment (zeroFruit) [inline conn.go]: `Ip` should return ip not member id!

**Review by zeroFruit — APPROVED** [review]

**Pull request merged**